File: internal/translate/ir.go
package translate

import "encoding/json"

// Request is the normalized, format-agnostic representation of an inbound
// chat completion request.
type Request struct {
	Model       string
	System      []TextBlock
	Messages    []Message
	Tools       []Tool
	ToolChoice  *ToolChoice
	MaxTokens   int
	Temperature *float64
	TopP        *float64
	Stream      bool
	Stop        []string
	Extra       map[string]any
}

type TextBlock struct {
	Text string
}

type Message struct {
	Role    string // "user" | "assistant"
	Content []ContentBlock
}

// ContentBlock is a discriminated union; Type selects the populated field.
type ContentBlock struct {
	Type       string // "text" | "image" | "tool_use" | "tool_result"
	Text       string
	Image      *Image
	ToolUse    *ToolUse
	ToolResult *ToolResult
}

type Image struct {
	URL       string // http(s) URL (OpenAI image_url form)
	Base64    string // base64-encoded data (Anthropic source form)
	MediaType string // media type when Base64 is set
}

type ToolUse struct {
	ID    string
	Name  string
	Input json.RawMessage
}

type ToolResult struct {
	ToolUseID string
	Content   []ContentBlock
	IsError   bool
}

type Tool struct {
	Name        string
	Description string
	InputSchema json.RawMessage
}

type ToolChoice struct {
	Type string // "auto" | "none" | "tool"
	Name string // set when Type == "tool"
}

type Response struct {
	ID         string
	Model      string
	Content    []ContentBlock
	StopReason string // canonical IR vocabulary: "stop" | "max_tokens" | "tool_calls" | "content_filter"
	Usage      Usage
	Extra      map[string]any
}

type Usage struct {
	InputTokens  int
	OutputTokens int
}

// StreamEvent is an Anthropic-style fine-grained streaming event.
type StreamEvent struct {
	Type         string        // message_start | content_block_start | content_block_delta | content_block_stop | message_delta | message_stop
	MessageID    string        // message_start
	Model        string        // message_start
	InputTokens  int           // message_start
	Index        int           // content_block_*
	Block        *ContentBlock // content_block_start
	Delta        *Delta        // content_block_delta
	StopReason   string        // message_delta
	OutputTokens int           // message_delta
}

type Delta struct {
	Type        string // "text_delta" | "input_json_delta"
	Text        string // text_delta
	PartialJSON string // input_json_delta
}

File: web/src/utils/ir.ts
/**
 * 归档的 IR（中间表示）JSON 结构 —— Go 默认序列化，字段为 PascalCase。
 * 未设置的字段可能缺失或为零值（空字符串 / null），须按 Type 判断内容块种类。
 */

export interface IRSystemBlock {
  Text: string
}

export interface IRImage {
  URL: string
  Base64: string
  MediaType: string
}

export interface IRToolUse {
  ID: string
  Name: string
  /** json.RawMessage —— 可能是对象，也可能是字符串 */
  Input: unknown
}

export interface IRToolResult {
  ToolUseID: string
  Content: unknown
  IsError: boolean
}

export type IRContentType = 'text' | 'image' | 'tool_use' | 'tool_result' | 'thinking' | 'redacted_thinking'

export interface IRContentBlock {
  Type: IRContentType
  Text: string
  Thinking: string
  Signature: string
  Data: string
  Image: IRImage | null
  ToolUse: IRToolUse | null
  ToolResult: IRToolResult | null
}

export interface IRMessage {
  Role: 'user' | 'assistant'
  Content: IRContentBlock[]
}

export interface IRRequest {
  Model: string
  System?: IRSystemBlock[]
  Messages?: IRMessage[]
  Tools?: unknown[]
}

/** 对应 Go translate.Usage（无 json tag，PascalCase 序列化） */
export interface IRUsage {
  InputTokens?: number
  OutputTokens?: number
  CacheReadTokens?: number
  CacheCreationTokens?: number
  ReasoningTokens?: number
}

export interface IRResponse {
  ID: string
  Model: string
  Content?: IRContentBlock[]
  StopReason: string
  Usage?: IRUsage
}

/** 安全解析 IR JSON 文本；非法 JSON / 空串返回 null */
export function parseIR<T>(json: string): T | null {
  if (!json) return null
  try {
    return JSON.parse(json) as T
  } catch {
    return null
  }
}
